src/colmap/estimators/bundle_adjustment.cc

```
@@ -540,23 +639,110 @@ class DefaultBundleAdjuster : public BundleAdjuster {
-        }
       }
     }
   }
 
+  void AddImageWithNonTrivialFrame(Image& image,
+                                   Reconstruction& reconstruction) {
+    Camera& camera = *image.CameraPtr();
+
+    THROW_CHECK(!image.HasTrivialFrame());
+    Rigid3d& cam_from_rig = image.FramePtr()->RigPtr()->SensorFromRig(
+        sensor_t(SensorType::CAMERA, image.CameraId()));
+    Rigid3d& frame_from_world = image.FramePtr()->FrameFromWorld();
+
+    // CostFunction assumes unit quaternions.
+    cam_from_rig.rotation.normalize();
+    frame_from_world.rotation.normalize();
+
+    double* cam_from_rig_rotation = cam_from_rig.rotation.coeffs().data();
+    double* cam_from_rig_translation = cam_from_rig.translation.data();
+    double* frame_from_world_rotation =
+        frame_from_world.rotation.coeffs().data();
+    double* frame_from_world_translation = frame_from_world.translation.data();
+    double* camera_params = camera.params.data();
+
+    const bool constant_sensor_from_rig =
+        !options_.refine_sensor_from_rig ||
+        config_.HasConstantSensorFromRigPose(image.CameraPtr()->SensorId());
+    const bool constant_frame_from_world =
+        !options_.refine_frame_from_world ||
+        config_.HasConstantFrameFromWorldPose(image.FrameId());
+    const bool constant_pose_cost_functor =
+        constant_frame_from_world && constant_sensor_from_rig;
+    const bool constant_rig_cost_functor =
+        !constant_frame_from_world && constant_sensor_from_rig;
+
+    // Add residuals to bundle adjustment problem.
+    size_t num_observations = 0;
+    for (const Point2D& point2D : image.Points2D()) {
+      if (!point2D.HasPoint3D()) {
+        continue;
+      }
+
+      num_observations += 1;
+      point3D_num_observations_[point2D.point3D_id] += 1;
+
+      Point3D& point3D = reconstruction.Point3D(point2D.point3D_id);
+      THROW_CHECK_GT(point3D.track.Length(), 1);
+
+      // Notice that the constant_frame_from_world && !constant_sensor_from_rig
```

Again, should we throw an error when the rare/exotic case happens?

---

I don't think it makes sense to throw an exception. It all works as expected - it's just a little less efficient as with a customized cost function for the special case.